internal/restconfig/restconfig.go

```diff
@@ -52,24 +54,112 @@ type RestConfig struct {
+	namespaceFlag       bool
 }
 
-func NewProducer() Producer {
-	return Producer{}
+// NewProducer initialises a blank producer which needs to be set up with flags (see SetupFlags).
+func NewProducer() *Producer {
+	return &Producer{}
 }
 
-func NewProducerFrom(kubeConfig, kubeContext string) Producer {
-	return Producer{
+// NewProducerFrom initialises a producer using the given kubeconfig file and context.
+// The context may be empty, in which case the default context will be used.
+func NewProducerFrom(kubeConfig, kubeContext string) *Producer {
+	return &Producer{
 		kubeConfig:  kubeConfig,
 		kubeContext: kubeContext,
 	}
 }
 
+// WithNamespace configures the producer to set up a namespace flag.
+// The chosen namespace will be passed to the PerContextFn used to process the context.
+func (rcp *Producer) WithNamespace() *Producer {
+	rcp.namespaceFlag = true
+
+	return rcp
+}
+
+// SetupFlags configures the given flags to control the producer settings.
+func (rcp *Producer) SetupFlags(flags *pflag.FlagSet) {
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	loadingRules.DefaultClientConfig = &clientcmd.DefaultClientConfig
+
+	flags.StringVar(&loadingRules.ExplicitPath, "kubeconfig", "", "absolute path(s) to the kubeconfig file(s)")
+
+	// Default un-prefixed context
+	overrides := clientcmd.ConfigOverrides{ClusterDefaults: clientcmd.ClusterDefaults}
+	kflags := clientcmd.RecommendedConfigOverrideFlags("")
+	clientcmd.BindOverrideFlags(&overrides, flags, kflags)
+
+	// Support deprecated --kubecontext; TODO remove in 0.15
+	legacyFlags := clientcmd.RecommendedConfigOverrideFlags("kube")
+	legacyFlags.CurrentContext.BindStringFlag(flags, &overrides.CurrentContext)
+	_ = flags.MarkDeprecated("kubecontext", "use --context instead")
+
+	if !rcp.namespaceFlag {
+		_ = flags.MarkHidden("namespace")
+	}
+
+	rcp.defaultClientConfig = &configAndOverrides{
+		config:    clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, &overrides),
+		overrides: &overrides,
+	}
+}
+
+type PerContextFn func(clusterInfo *cluster.Info, namespace string, status reporter.Interface) error
+
+// RunOnSelectedContext runs the given function on the selected context.
+func (rcp *Producer) RunOnSelectedContext(function PerContextFn, status reporter.Interface) error {
+	if rcp.inCluster {
+		restConfig, err := rest.InClusterConfig()
+		if err != nil {
+			return status.Error(err, "error retrieving the in-cluster configuration")
+		}
+
+		// In-cluster configurations don't give a cluster name, use "in-cluster"
+		clusterInfo, err := cluster.NewInfo("in-cluster", restConfig)
+		if err != nil {
+			return status.Error(err, "error building the cluster.Info for the in-cluster configuration")
+		}
+
+		// In-cluster configurations don't specify a namespace, use the default
+		// When using the in-cluster configuration, that's the only configuration we want
+		return function(clusterInfo, "default", status)
+	}
+
+	if rcp.defaultClientConfig != nil {
```

Comment:
Shouldn't it error out if the config is not properly initialized?

I feel like we might miss a bug otherwise, as it'll just return no error but also not do anything AFAICT

Reply:
Ah yes, I meant to address that earlier and forgot, thanks for noticing!

Reply:
Thanks, it's usually better (i.e. makes the code more maintainable) to address the exceptional case in the if, not the normal flow, so I'd expect it to look something like this:
```suggestion
	// If we get here, no context was set up, which means SetupFlags() wasn't called
	if rcp.defaultClientConfig == nil {
		return status.Error(errors.New("no context provided (this is a programming error)"), "")
	}
```